### PR TITLE
Task/jenkins 38150 Handle declarative pipeline preparatory steps and postBuild stage

### PR DESCRIPTION
**Pull request by vivek**
# Description

See [JENKINS-38150](https://issues.jenkins-ci.org/browse/JENKINS-38150).

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
There are snapshot dependencies on pipeline-model-definition and workflow-api.
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by vivek** [comment]
@i386, @scherler  There is side effect of marking skipped stage's state as SKIPPED and result as NOT_BUILT. It shows up as purple with question mark. See the attached image.

image

**Comment by i386** [comment]
@vivek did you notice any issues with karaoke? 

I've opened up [JENKINS-39871](https://issues.jenkins-ci.org/browse/JENKINS-39871) to track updating the status on the pipeline graph.

**Comment by vivek** [comment]
@i386 Karaoke looked fine.

**Comment by i386** [comment]
I'd like to give this a week to soak on dogfood. 

Could we rewrite the Jenkinsfile for Blue Ocean in this PR using Declarative as validation?

**Comment by vivek** [comment]
@i386 here is the pipeline script I used for some of the testing, https://gist.github.com/vivek/e1a07935e7f70ca4ca1766db1ef0fbc9. To get preparatory stage, its easy to use a Jenkinsfile on local git repo and the create a pipeline job using that git repo.

And yeah we should convert blueocean jenkinsfile using declarative.

**Comment by michaelneale** [comment]
We depend on dogfood quite a lot and the PMD plugin is still undergoing changes, so I would be reluctant to switch for master just yet - but a long running branch with it would be fine. 

For this to "soak" on dogfood it needs to be merged to master, however?

**Comment by i386** [comment]
There is no release this week so this can be merged.

**Comment by michaelneale** [comment]
LGTM - I am trying the ATH with the branch that has the 0.6 version of PMD pipelines: 

https://ci.blueocean.io/job/ATH-Jenkinsfile/job/declarative-0.6-upgrade/3/

so if that passes, why not :bee: (I am assuming that branch of ATH still applies here). 
This is a little out of date with master, so may want to have it up to date if not 100% sure.

**Comment by michaelneale** [comment]
ATH says :shipit: - nice.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 1]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+        }
+        return null;
+    }
+
+    public static boolean isPostSyntheticStage(@Nonnull FlowNode node){
+        TagsAction tagsAction = getSyntheticStage(node);
+        if(tagsAction == null){
+            return false;
+        }
+        String value = tagsAction.getTagValue("SYNTHETIC_STAGE");
```
Comment (jglick) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: Magic constants like these needed to be defined upstream.

Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: It is - https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/47/files#diff-2582f0481d4a2304989407f7fd13dea1R37

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 2]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
-    public static boolean isParallelBranch(FlowNode node){
+    public static boolean isSyntheticStage(@Nullable FlowNode node){
+        return node!= null && getSyntheticStage(node) != null;
+    }
+
+    @CheckForNull
+    public static TagsAction getSyntheticStage(@Nullable FlowNode node){
+        if(node != null) {
+            for (Action action : node.getActions()) {
+                if (action instanceof TagsAction && ((TagsAction) action).getTagValue("SYNTHETIC_STAGE") != null) {
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: Should be using `org.jenkinsci.plugins.pipeline.SyntheticStage.getTagName()` - I'll make an easier way to get the appropriate `StageTagsMetadata` extension for a given tag name.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java (hunk 1 above)
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: See above.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 3]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+        String value = tagsAction.getTagValue("SYNTHETIC_STAGE");
+        return value!=null && value.equals("POST");
+    }
+
+    public static boolean isSkippedStage(@Nullable FlowNode node){
+        if(node == null){
+            return false;
+        }
+        for (Action action : node.getActions()) {
+            if (action instanceof TagsAction && ((TagsAction) action).getTagValue("STAGE_STATUS") != null) {
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: `StageStatus.getTagName()` - see above.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 4]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+                TagsAction tagsAction =  (TagsAction) action;
+                String value = tagsAction.getTagValue("STAGE_STATUS");
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: See above.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 5]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+                return value != null && value.equals("SKIPPED_FOR_CONDITIONAL");
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: `StageStatus.getSkippedForConditional()` - see above.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 6]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+        return false;
+    }
+
+    public static boolean isPreSyntheticStage(@Nonnull FlowNode node){
+        TagsAction tagsAction = getSyntheticStage(node);
+        if(tagsAction == null){
+            return false;
+        }
+        String value = tagsAction.getTagValue("SYNTHETIC_STAGE");
+        return value!=null && value.equals("PRE");
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: `SyntheticStage.getPre()` - see above

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java [hunk 7]
```diff
@@ -59,19 +66,68 @@ public static NodeRunStatus getStatus(WorkflowRun run){
...
+        return null;
+    }
+
+    public static boolean isPostSyntheticStage(@Nonnull FlowNode node){
+        TagsAction tagsAction = getSyntheticStage(node);
+        if(tagsAction == null){
+            return false;
+        }
+        String value = tagsAction.getTagValue("SYNTHETIC_STAGE");
+        return value!=null && value.equals("POST");
```
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: `SyntheticStage.getPost()` - see above.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java (hunk 1 above)
Comment (vivek) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: @abayer I would expect just static of enum of these constants? Is this defined somewhere?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java (hunk 2 above)
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: I take that back - `ExtensionList.lookup(StageTagsMetadata.class).get(SyntheticStage.class)` does the trick, no new code needed.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java (hunk 1 above)
Comment (abayer) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeUtil.java]: I thought about that but went in an `Extension` direction to make it more, well, extensible. =)

**Comment by michaelneale** [comment]
thanks @abayer. I guess this is still a WIP @vivek ?

**Comment by vivek** [comment]
@abayer SyntheticStage is Extension, that is its singleton. From consumer of your API, I will always use this singleton to get the constant - this seems overkill. In effect its equivalent to using pre-defined constant (as public static final or enum kind of constant). Or maybe you think these constants are going to change in the future?

**Comment by abayer** [comment]
@vivek So I went with the `Extension` so that (a) it's extensible in the future, and (b) I'm not just cobbling together various `public static final`s to expose this information. It just felt a lot less awkward to me, but I can be persuaded. =)

**Comment by vivek** [comment]
@abayer If I was using constant as in enum or public constant for SyntheticStage or StageStatus, then for one it avoids checking for null, which is bit ugly and anyone who is going to use this API has to have null check as 'ExtensionList.lookup(StageTagsMetadata.class).get(SyntheticStage.class)' could be null vs SyntheticStage.SYNTHETIC_STAGE_TAG or SyntheticStage.PRE_TAG etc.

**Comment by abayer** [comment]
@vivek Ok. Hrm. I just don't know how I can avoid a mess of `public static final`s for all the possible fields...let me think about this a bit.

**Comment by vivek** [comment]
@abayer To be clear, SyntheticStage or StageStatus implemented as Extension of StageTagsMetadata is good design. What I am pointing is that just to pick constant names of tags, I should not be forced to pick these Extensions, they could be defined as static constants in SyntheticStage or StageStatus. SyntheticStage or StageStatus are themselves singletons, and tags they expose is a constant in implementation so why not expose them as constant?

**Comment by abayer** [comment]
@vivek Well, there's literally nothing to them right now but tag names and possible values currently. =)

**Comment by abayer** [comment]
@vivek new commit exposes things like `SyntheticStage.TAG_NAME`, `SyntheticStage.getPre()` etc as statics.

**Comment by vivek** [comment]
@abayer thanks!

**Comment by vivek** [comment]
@abayer pushed commit that uses constants defined in declarative plugin.

**Comment by michaelneale** [comment]
@vivek not sure what those sudden conflicts are though

**Comment by vivek** [comment]
@michaelneale other commits on master had conflict with this branch. Fixed.

File: blueocean/pom.xml [hunk 8]
```diff
@@ -54,13 +54,6 @@
         </dependency>
 
         <!-- Bundled plugins for improved out of the box experience -->
-        <!-- Kyoto feature -->
-        <dependency>
```
Comment (michaelneale) [inline blueocean/pom.xml]: so this is removed from here?

**Comment by michaelneale** [comment]
if it passes ATH - I guess that is good - what should I be reviewing - is there a good sample to try? code is fine.

**Comment by cliffmeyers** [comment]
@michaelneale you can take a quick peek at the new Dialog if you like.

**Comment by michaelneale** [comment]
:bee: assuming ATH, and upstream published dependencies etc - but seems to work as advertised for a quick test and andrew likes the look of it.

**Comment by i386** [comment]
@abayer LGTY?

**Comment by vivek** [comment]
@michaelneale @abayer @i386 upgraded to 0.7.1 declarative and 2.7 workflow-api and merged with latest master. Please review.

**Comment by michaelneale** [comment]
:bee: from me and passes ATH 

so depends if ou want to see if @abayer has a look

File: blueocean/pom.xml (hunk 8 above)
Comment (michaelneale) [inline blueocean/pom.xml]: ok

**Comment by vivek** [comment]
@abayer please confirm if declarative version and workflow-api versions are correct ones, I think they are but will wait for your confirmation before merge.

**Comment by abayer** [comment]
:bee: lgtm

**Pull request merged**